src/Glide64/3dmath.cpp

```diff
@@ -190,16 +190,37 @@ void InverseTransformVectorC (float *src, float *dst, float mat[4][4])
 
 void MulMatricesC(float m1[4][4],float m2[4][4],float r[4][4])
 {
-  for (int i=0; i<4; i++)
-  {
-    for (int j=0; j<4; j++)
+    float row[4][4];
+    register unsigned int i, j;
+
+    for (i = 0; i < 4; i++)
+        for (j = 0; j < 4; j++)
+            row[i][j] = m2[i][j];
+    for (i = 0; i < 4; i++)
     {
-      r[i][j] = m1[i][0] * m2[0][j] +
-                m1[i][1] * m2[1][j] +
-                m1[i][2] * m2[2][j] +
-                m1[i][3] * m2[3][j];
+        float leftrow[4], destrow[4];
```

Comment:
I just tested and yeah, GCC doesn't like nested loops, even small ones...

I just tested on a old gcc (4.4 64-bits):
```c
for (j = 0; j < 4; j++)
    summand[0][j] = leftrow[0] * row[0][j];
for (j = 0; j < 4; j++)
    summand[1][j] = leftrow[1] * row[1][j];
for (j = 0; j < 4; j++)
    summand[2][j] = leftrow[2] * row[2][j];
for (j = 0; j < 4; j++)
    summand[3][j] = leftrow[3] * row[3][j];
```
Result to this:
```asm
	movss	leftrow(%rip), %xmm0
	xorl	%eax, %eax
	movss	row(%rip), %xmm1
	mulss	%xmm0, %xmm1
	movss	%xmm1, summand(%rip)
	movss	row+4(%rip), %xmm1
	mulss	%xmm0, %xmm1
	movss	%xmm1, summand+4(%rip)
	movss	row+8(%rip), %xmm1
	mulss	%xmm0, %xmm1
	mulss	row+12(%rip), %xmm0
	movss	%xmm1, summand+8(%rip)
	movss	row+16(%rip), %xmm1
	movss	%xmm0, summand+12(%rip)
	movss	leftrow+4(%rip), %xmm0
	mulss	%xmm0, %xmm1
	movss	%xmm1, summand+16(%rip)
	movss	row+20(%rip), %xmm1
	mulss	%xmm0, %xmm1
	movss	%xmm1, summand+20(%rip)
	movss	row+24(%rip), %xmm1
	mulss	%xmm0, %xmm1
	mulss	row+28(%rip), %xmm0
	movss	%xmm1, summand+24(%rip)
	movss	row+32(%rip), %xmm1
	movss	%xmm0, summand+28(%rip)
	movss	leftrow+8(%rip), %xmm0
	mulss	%xmm0, %xmm1
	movss	%xmm1, summand+32(%rip)
	movss	row+36(%rip), %xmm1
	mulss	%xmm0, %xmm1
	movss	%xmm1, summand+36(%rip)
	movss	row+40(%rip), %xmm1
	mulss	%xmm0, %xmm1
	mulss	row+44(%rip), %xmm0
	movss	%xmm1, summand+40(%rip)
	movss	row+48(%rip), %xmm1
	movss	%xmm0, summand+44(%rip)
	movss	leftrow+12(%rip), %xmm0
	mulss	%xmm0, %xmm1
	movss	%xmm1, summand+48(%rip)
	movss	row+52(%rip), %xmm1
	mulss	%xmm0, %xmm1
	movss	%xmm1, summand+52(%rip)
	movss	row+56(%rip), %xmm1
	mulss	%xmm0, %xmm1
	movss	%xmm1, summand+56(%rip)
	mulss	row+60(%rip), %xmm0
	movss	%xmm0, summand+60(%rip)
```
I'm not confortable with asm but this result seems "silly". Is it me?

This:
```c
for (j = 0; j < 4; j++){
    summand[0][j] = leftrow[0] * row[0][j];
    summand[1][j] = leftrow[1] * row[1][j];
    summand[2][j] = leftrow[2] * row[2][j];
    summand[3][j] = leftrow[3] * row[3][j];
}
```
Result to this (your result):
```asm
	movss	leftrow(%rip), %xmm0
	xorl	%eax, %eax
	shufps	$0, %xmm0, %xmm0
	mulps	row(%rip), %xmm0
	movaps	%xmm0, summand(%rip)
	movss	leftrow+4(%rip), %xmm0
	shufps	$0, %xmm0, %xmm0
	mulps	row+16(%rip), %xmm0
	movaps	%xmm0, summand+16(%rip)
	movss	leftrow+8(%rip), %xmm0
	shufps	$0, %xmm0, %xmm0
	mulps	row+32(%rip), %xmm0
	movaps	%xmm0, summand+32(%rip)
	movss	leftrow+12(%rip), %xmm0
	shufps	$0, %xmm0, %xmm0
	mulps	row+48(%rip), %xmm0
	movaps	%xmm0, summand+48(%rip)
```
And:
```c
    for (k = 0; k < 4; k++)
        for (j = 0; j < 4; j++)
            summand[k][j] = leftrow[k] * row[k][j];
```
Is the worse:
```asm
	movaps	row(%rip), %xmm0
	xorl	%eax, %eax
	movaps	row+16(%rip), %xmm3
	movaps	%xmm0, %xmm2
	movaps	row+32(%rip), %xmm5
	shufps	$221, %xmm3, %xmm0
	shufps	$136, %xmm3, %xmm2
	movaps	row+48(%rip), %xmm1
	movaps	%xmm5, %xmm6
	shufps	$136, %xmm1, %xmm6
	movaps	%xmm2, %xmm3
	shufps	$221, %xmm1, %xmm5
	movaps	%xmm0, %xmm1
	shufps	$136, %xmm6, %xmm3
	movaps	leftrow(%rip), %xmm4
	shufps	$136, %xmm5, %xmm1
	shufps	$221, %xmm6, %xmm2
	mulps	%xmm4, %xmm3
	shufps	$221, %xmm5, %xmm0
	mulps	%xmm4, %xmm2
	mulps	%xmm4, %xmm1
	mulps	%xmm4, %xmm0
	movaps	%xmm3, %xmm4
	unpckhps	%xmm2, %xmm3
	unpcklps	%xmm2, %xmm4
	movaps	%xmm1, %xmm2
	unpcklps	%xmm0, %xmm2
	unpckhps	%xmm0, %xmm1
	movaps	%xmm4, %xmm0
	unpcklps	%xmm2, %xmm0
	unpckhps	%xmm2, %xmm4
	movaps	%xmm0, summand(%rip)
	movaps	%xmm3, %xmm0
	unpckhps	%xmm1, %xmm3
	unpcklps	%xmm1, %xmm0
	movaps	%xmm4, summand+16(%rip)
	movaps	%xmm0, summand+32(%rip)
	movaps	%xmm3, summand+48(%rip)
```
So, on gcc 4.4, option 2 seems to be the best.

I'm impressed how results can change in this simple loop doing the exact same thing...